app/src/hid_keyboard.h

```diff
@@ -0,0 +1,89 @@
+#ifndef HID_KEYBOARD_H
+#define HID_KEYBOARD_H
+
+#include <stdbool.h>
+
+#include <SDL2/SDL.h>
+
+#include "aoa_hid.h"
+
+/**
+ * Because of dual-report, when we send hid events, we need to add report id
+ * as prefix, so keyboard keys looks like
+ * `0x01 Modifier Reserved Key Key Key Key Key Key` and media keys looks like
+ * `0x02 MediaMask` (one key per bit for media keys).
+ */
+#define HID_REPORT_ID_INDEX 0
+#define HID_KEYBOARD_MODIFIER_INDEX (HID_REPORT_ID_INDEX + 1)
+#define HID_KEYBOARD_MODIFIER_NONE 0x00
+#define HID_KEYBOARD_MODIFIER_LEFT_CONTROL (1 << 0)
+#define HID_KEYBOARD_MODIFIER_LEFT_SHIFT (1 << 1)
+#define HID_KEYBOARD_MODIFIER_LEFT_ALT (1 << 2)
+#define HID_KEYBOARD_MODIFIER_LEFT_GUI (1 << 3)
+#define HID_KEYBOARD_MODIFIER_RIGHT_CONTROL (1 << 4)
+#define HID_KEYBOARD_MODIFIER_RIGHT_SHIFT (1 << 5)
+#define HID_KEYBOARD_MODIFIER_RIGHT_ALT (1 << 6)
+#define HID_KEYBOARD_MODIFIER_RIGHT_GUI (1 << 7)
+// USB HID protocol says 6 keys in an event is the requirement for BIOS
+// keyboard support, though OS could support more keys via modifying the report
+// desc, I think 6 is enough for us.
+#define HID_KEYBOARD_MAX_KEYS 6
+#define HID_KEYBOARD_EVENT_SIZE (3 + HID_KEYBOARD_MAX_KEYS)
+#define HID_KEYBOARD_REPORT_ID 0x01
+#define HID_KEYBOARD_RESERVED 0x00
+#define HID_KEYBOARD_ERROR_ROLL_OVER 0x01
+#define HID_MEDIA_EVENT_SIZE 2
+#define HID_MEDIA_REPORT_ID 0x02
+#define HID_MEDIA_KEY_INDEX (HID_REPORT_ID_INDEX + 1)
+/**
+ * Media keys handle as mask so we define them here.
+ * Currently not used because desktop environment catches them before window.
```

Comment:
I think that we should not forward media keys anyway (which are "global shortcuts"): if I want to change volume or mute my desktop, I don't want to also mute or change volume on the device as a consequence.

So it's a good thing that the desktop environment catches them. I think we could remove these defines.

Reply:
Oh, unless they are generated from the scrcpy shortcuts (e.g. Alt + up could trigger the HID media volume up). :+1: